### PR TITLE
feat!: migrate runtime, plugin API, and CLI to Effect

### PR DESCRIPTION
**Pull request by typeweaver**
## Summary

Completes TypeWeaver's migration to an Effect-based runtime, plugin architecture, and CLI while preserving deterministic generated output.

- **Effect-native plugin API** — replaces `BasePlugin` classes with records created through `definePlugin(...)` / `definePluginWithLibCopy(...)`. Lifecycle hooks return `Effect`, and all first-party plugins use the new contract.
- **Effect service architecture** — generation, loading, registry, configuration, spec processing, formatting, path safety, template rendering, and platform I/O are expressed as composable services and layers with typed failures.
- **Effect CLI** — replaces Commander with `@effect/cli`, keeps the public `-V` version alias, provides concise tagged-error output, and enables debug logs through `--verbose`.
- **Safety and concurrency** — provides per-call registry isolation, ownership-aware output locks, interruption-safe reverse finalization, atomic writes, strict generated-path checks, and conservative orphan cleanup.
- **Plugin authoring surface** — offers synchronous compatibility helpers and Effect-native context helpers, validates imported plugin/config shapes, and verifies packed external consumers against the supported Effect peer range.
- **Maintainability** — enforces the requested complexity limits through Oxlint/`eslint-plugin-sonarjs` compatibility without installing ESLint.

## Breaking changes

- V1 `BasePlugin` subclasses must migrate to `definePlugin(...)` records.
- Plugin lifecycle hooks return `Effect`; typed `finalize` failures are logged as warnings, while defects and interruptions remain observable.
- Tagged error constructors use object arguments.
- Plugin registry consumers create an isolated instance per generation call.
- CLI presentation and error formatting differ from Commander, although command/flag compatibility is covered by process tests.
- Published packages accept `effect >=3.22.0 <4`; the workspace and test matrix use Effect 3.22.

The spec-authoring API and Zod schemas remain unchanged. Generated fixtures are checked byte-for-byte.

## Documentation

- [Migration guide: 0.12.x to 1.0.x](MIGRATION.md#migrating-from-012x-to-10x)
- [Plugin authoring guide](docs/plugin-authoring.md)
- [ADR 0003 — Effect-native plugin API](docs/adr/0003-effect-native-plugin-api.md)
- [ADR 0004 — FileSystem service adoption](docs/adr/0004-filesystem-service-adoption.md)
- [ADR 0005 — Effect service patterns](docs/adr/0005-effect-service-patterns.md)
- [ADR 0006 — CLI error and log formatting](docs/adr/0006-cli-error-and-log-formatting.md)
- [ADR 0007 — Generator per-call isolation](docs/adr/0007-generator-per-call-isolation.md)
- [ADR 0008 — Effect version discipline](docs/adr/0008-effect-version-discipline.md)

## Verification

- [x] Frozen pnpm install
- [x] Actionlint
- [x] Oxfmt and Oxlint, including maintainability limits
- [x] Strict workspace typecheck
- [x] All-package build and size limits
- [x] 2,379 workspace tests (plus two intentional platform-specific skips)
- [x] 225 generated fixtures match exactly
- [x] Effect migration diagnostics report zero findings
- [x] Packed external consumers install and execute with Effect 3.22
- [x] Duplicate Effect identity is rejected
- [x] Built CLI process contracts pass on Linux and Windows
- [x] Windows security/path/locking contracts pass
- [x] CodeQL and Socket security gates run on the exact PR head

The reproducible umbrella command is:

```sh
pnpm verify:effect-migration
```

### CODE REVIEW

**Comment by typeweaver** [comment]
@codex review

**Comment by typeweaver** [comment]
@codex review

File: packages/cli/package.json [hunk 1]
```diff
@@ -72,6 +75,17 @@
   },
   "homepage": "https://github.com/rexeus/typeweaver#readme",
   "dependencies": {
+    "@effect/cli": "^0.76.0",
+    "@effect/cluster": "^0.60.0",
+    "@effect/experimental": "^0.61.0",
```
[the remark this answers is not in the document]
Comment (typeweaver) [reply packages/cli/package.json:80]: Verified against the installed package metadata and published bundle: these are not unused optional packages. `@effect/platform-node@0.108.0` has non-optional peers on `@effect/cluster`, `@effect/rpc`, and `@effect/sql`; cluster requires workflow; sql/workflow require experimental. `@effect/cli` requires printer/printer-ansi, whose non-optional peer is typeclass. The CLI build externalizes these packages, and this repository deliberately sets `auto-install-peers=false`, so removing or moving them to devDependencies would publish an incomplete runtime dependency graph. `pnpm why` confirms each chain. Reducing this graph would require replacing the platform-node/CLI stack, not deleting these declarations.

File: packages/gen/src/services/internal/pluginContextBuilder.ts [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+
+    if (existingFileMode !== undefined) {
+      fileSystem.chmod(tempFile, existingFileMode);
+    }
+
+    // Both operations are synchronous: once rename publishes the destination,
+    // record the write before any fallible cleanup can run. A cleanup-only
+    // failure may still be reported, but it cannot leave a committed file
+    // absent from the generated-file tracker and pending log queue.
+    fileSystem.rename(tempFile, config.safePath.fullPath);
```
[the remark this answers is not in the document]
Comment (typeweaver) [reply packages/gen/src/services/internal/pluginContextBuilder.ts]: Fixed in 747d5461. Both sync and Effect-native writers now capture a stable absolute output root, validate before directory creation, revalidate before staging, and revalidate again immediately before rename. Typed path failures prevent rename/tracking and scoped cleanup still runs. Tests cover the typed Effect failure/cleanup path and a real deterministic ancestor symlink swap proving no rename, no tracking, and no write outside outputDir. The relevant suites are now included in the Windows security gate.

File: packages/cli/src/services/generatorIO.ts [hunk 3]
```diff
@@ -0,0 +1,613 @@
...
+    if (!entry.isDirectory()) {
+      continue;
+    }
+    if (isOutputLockArtifact(entry.name)) {
+      continue;
+    }
+    const entryPath = path.join(directory, entry.name);
+    if (isAtomicWriteTempDirectory(entry.name)) {
+      fs.rmSync(entryPath, { recursive: true, force: true });
+      continue;
```
[the remark this answers is not in the document]
Comment (typeweaver) [reply packages/cli/src/services/generatorIO.ts]: Fixed in 8127dc06. The orphan sweep now recognizes the two exact Node mkdtemp ownership shapes: `.typeweaver-` plus six alphanumeric characters and `.typeweaver-spec-loader-` plus six. A full `--no-clean` generation test seeds real mkdtemp crash debris under `spec/`, proves it is removed, proves final output is generated, and proves a similarly named user-owned directory survives. This suite is now part of the Windows security gate.

**Comment by typeweaver** [comment]
@codex review

**Comment by typeweaver** [comment]
@codex review

File: packages/cli/src/services/generatorIO.ts [hunk 4]
```diff
@@ -0,0 +1,622 @@
...
+ * channel, so the strict operation is deliberately downgraded to a warning
+ * here rather than masking the pipeline's own outcome.
+ */
+export const releaseOutputLock = (lock: OutputLock): Effect.Effect<void> =>
+  releaseOutputLockStrict(lock).pipe(
+    Effect.catchAll(failure =>
+      Effect.logWarning(
+        `Failed to release output lock at '${lock.path}': ${failure.message}`
+      )
+    )
```
[the remark this answers is not in the document]
Comment (typeweaver) [reply packages/cli/src/services/generatorIO.ts]: Fixed in e664d4c7. Lock release now token-checks and atomically detaches the canonical lock into its ownership fence before best-effort cleanup. Typed metadata-read or detach failures remember only the exact owner token, allowing a repaired long-lived runtime to reclaim its own abandoned live-PID lock while active same-PID and replacement owners remain contention. Deterministic tests cover metadata-read failure, detach failure, partial fence cleanup, and replacement isolation.

File: packages/cli/src/services/cleanTargetGuard.ts [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+  inputFile: string | undefined
+): void => {
+  if (inputFile === undefined) {
+    return;
+  }
+  const canonicalInputFile = canonicalizePathForContainment(
+    path.resolve(context.resolvedWorkingDirectory, inputFile),
+    context.fileSystem
+  );
+  if (!isSameOrDescendantOf(canonicalInputFile, context.canonicalOutputDir)) {
```
[the remark this answers is not in the document]
Comment (typeweaver) [reply packages/cli/src/services/cleanTargetGuard.ts]: Fixed in e664d4c7. Clean containment now checks both the resolved lexical input path against the resolved output and the canonical input against the canonical output. A real directory-symlink regression proves the link and external source survive because rejection happens before cleaning.

File: packages/cli/src/services/generatorIO.ts [hunk 6]
```diff
@@ -0,0 +1,622 @@
...
+      }
+      const entries = fs.readdirSync(outputDir, { withFileTypes: true });
+      for (const entry of entries) {
+        if (isOutputLockArtifact(entry.name)) {
+          continue;
+        }
+        fs.rmSync(path.join(outputDir, entry.name), {
+          recursive: true,
+          force: true,
+        });
```
[the remark this answers is not in the document]
Comment (typeweaver) [reply packages/cli/src/services/generatorIO.ts:140]: Fixed in e664d4c7. Existing and dangling symlinked output roots are rejected with a typed UnsafeCleanTargetError before destructive traversal. The target is revalidated under the held generation lock immediately before orphan sweep or clean, including no-clean runs. Real filesystem regressions prove external sentinels and owned-looking tempdirs survive preflight-to-mutation root swaps.

**Comment by typeweaver** [comment]
@codex review

File: packages/cli/src/services/cleanTargetGuard.ts [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+  }
+
+  const resolvedWorkingDirectory = path.resolve(currentWorkingDirectory);
+  const canonicalWorkingDirectory = fileSystem.realPath(
+    resolvedWorkingDirectory
+  );
+  const resolvedOutputDir = path.resolve(
+    resolvedWorkingDirectory,
+    trimmedOutputDir
+  );
```
[the remark this answers is not in the document]
Comment (typeweaver) [reply packages/cli/src/services/cleanTargetGuard.ts]: Fixed in 67ac6074. trim is now used only to reject all-whitespace input; resolution and canonicalization use the exact original output path that production creates, locks, sweeps, and cleans. Real POSIX and Windows-compatible whitespace-path regressions prove an external sentinel survives.

File: packages/cli/src/services/generatorIO.ts [hunk 8]
```diff
@@ -0,0 +1,739 @@
...
+    if (!isAsciiAlphaNumeric(character)) {
+      return false;
+    }
+  }
+  return true;
+};
+
+const isOwnedTempDirectory = (entryName: string): boolean =>
+  hasNodeTempDirectoryShape(entryName, TEMP_DIR_PREFIX) ||
+  hasNodeTempDirectoryShape(entryName, SPEC_BUNDLER_TEMP_DIR_PREFIX);
```
[the remark this answers is not in the document]
Comment (typeweaver) [reply packages/cli/src/services/generatorIO.ts]: Fixed in 67ac6074. Node-mkdtemp name shape is now only a candidate. Atomic writers and SpecBundler write an exact versioned regular-file ownership marker before staging data; sweep deletes only when both name shape and marker kind/content match. Exact-shaped user directories, including symlink-marker cases, are preserved and tested.

File: packages/cli/src/services/Formatter.ts [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+
+    for (const content of contents) {
+      // Skip atomic-write tempdirs and the lockfile sentinel — both are
+      // hidden coordination artifacts, not user-facing output. Walking
+      // into them would re-read/rewrite in-flight content from another
+      // run (`.typeweaver-XXXX/generated.tmp`) or the live lockfile
+      // metadata (`.typeweaver-lock/info.json`).
+      if (content.startsWith(".typeweaver-")) {
+        continue;
+      }
```
[the remark this answers is not in the document]
Comment (typeweaver) [reply packages/cli/src/services/Formatter.ts]: Fixed in 67ac6074. Formatter skips only exact lock/fence names and name-plus-marker verified coordination directories. Prefix-only files and directories remain format-visible. Unit and full plugin-generation regressions prove .typeweaver-output.ts is formatted, including an exact-shaped user directory with a symlinked marker.

File: packages/cli/src/services/cleanTargetGuard.ts [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+  });
+};
+
+const assertNotAncestorOfWorkingDirectory = (
+  context: CleanTargetContext,
+  protectedWorkspaceRoots: readonly string[]
+): void => {
+  const isProtectedAncestor =
+    protectedWorkspaceRoots.length > 0 &&
+    (isSameOrDescendantOf(
```
[the remark this answers is not in the document]
Comment (typeweaver) [reply packages/cli/src/services/cleanTargetGuard.ts]: Fixed in 67ac6074. Lexical and canonical ancestors of the current working directory are now rejected independently of workspace-marker discovery. Real markerless lexical and symlinked canonical ancestor tests prove the no-clean orphan sweep does not touch their sentinels.

File: packages/cli/src/services/SpecBundler.ts [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+ * — the FileSystem service is async-Effect and cannot satisfy that call
+ * site without restructuring the entire bundler. The sync `fs.realpathSync`
+ * is acceptable here because it runs at bundle time on user-supplied input
+ * paths only.
+ */
+const resolveRealFilePath = (filePath: string): string => {
+  if (!fs.existsSync(filePath)) {
+    return filePath;
+  }
+  return fs.realpathSync.native(filePath);
```
[the remark this answers is not in the document]
Comment (typeweaver) [reply packages/cli/src/services/SpecBundler.ts]: Fixed in 67ac6074. Bundle path construction and realpath probing now run inside Effect.try and map expected synchronous failures to SpecBundleError. A deterministic exists-to-realpath ENOENT regression proves zero defects, no Rolldown invocation, and scoped staging cleanup.

**Comment by typeweaver** [comment]
@codex review

**Comment by typeweaver** [comment]
@codex review

File: .github/workflows/quality-check.yml [hunk 12]
```diff
@@ -129,8 +131,41 @@ jobs:
...
+        run: pnpm install --frozen-lockfile
+
+      - name: Build all packages
+        run: pnpm run build
+
+      - name: Rebuild symlinks after build
+        run: pnpm install --frozen-lockfile
+
+      - name: Run Windows security contracts
+        run: pnpm run test:windows-security
```
[the remark this answers is not in the document]
Comment (typeweaver) [reply .github/workflows/quality-check.yml:171]: The full suite is already a blocking part of the Ubuntu quality-check. Its Verify Effect migration step runs pnpm verify:effect-migration; scripts/verify-effect-migration.mjs includes the command labeled Workspace unit, integration, and process tests with args -r --workspace-concurrency=1 --if-present test. That recursively executes every package test script, including core, clients, server, Hono, OpenAPI, CLI, and gen, and fails the enclosing CI step on any nonzero result. Quality Check run 30178831792 passed that exact step on 6bc2105d. The Windows job is intentionally an additional platform-specific subset, not a replacement, so adding a separate pnpm test step would duplicate the same full suite in the same Ubuntu job.

**Comment by typeweaver** [comment]
@codex review

**Comment by typeweaver** [comment]
@codex review

**Comment by typeweaver** [comment]
@codex review
